File: ui/src/utils/searchDsl.ts
import type { Sample } from "../types/sample";
import { matchesFuzzySearch } from "./search";

type NumberRange = {
  min: number | null;
  max: number | null;
};

type SampleSearchDsl = {
  textTerms: string[];
  excludedTerms: string[];
  bpm: NumberRange | null;
  playbackType: string | null;
  instrumentType: string | null;
  key: string | null;
  tags: string[];
  excludedTags: string[];
  favorite: boolean | null;
};

export function parseSampleSearchDsl(query: string): SampleSearchDsl {
  const parsed: SampleSearchDsl = {
    textTerms: [],
    excludedTerms: [],
    bpm: null,
    playbackType: null,
    instrumentType: null,
    key: null,
    tags: [],
    excludedTags: [],
    favorite: null,
  };

  const tokens = tokenize(query);
  if (tokens.length === 0 && query.trim()) {
    pushText(parsed, false, query);
    return parsed;
  }

  for (const token of tokens) {
    const negated = token.startsWith("-") && token.length > 1;
    const body = negated ? token.slice(1) : token;
    const separator = body.indexOf(":");

    if (separator === -1) {
      pushText(parsed, negated, body);
      continue;
    }

    const field = normalizeField(body.slice(0, separator));
    const value = body.slice(separator + 1).trim();
    if (!value) {
      pushText(parsed, negated, body);
      continue;
    }

    switch (field) {
      case "bpm":
        if (!negated) parsed.bpm = parseNumberRange(value);
        break;
      case "type":
      case "playback":
      case "playbacktype":
      case "sampletype":
        if (!negated) parsed.playbackType = normalizePlaybackType(value);
        break;
      case "instrument":
      case "instrumenttype":
        if (!negated) parsed.instrumentType = normalizeValue(value);
        break;
      case "key":
        if (!negated) parsed.key = normalizeKey(value);
        break;
      case "tag":
      case "tags":
        pushTag(parsed, negated, value);
        break;
      case "favorite":
      case "favourite":
      case "fav":
        if (!negated) parsed.favorite = parseBoolean(value);
        break;
      default:
        pushText(parsed, negated, body);
        break;
    }
  }

  return parsed;
}

export function matchesSampleSearchDsl(query: string, sample: Sample, isFavorite = false) {
  const parsed = parseSampleSearchDsl(query);
  const targets = [sample.file_name, ...sample.tags];

  for (const term of parsed.textTerms) {
    if (!matchesFuzzySearch(term, targets)) return false;
  }

  for (const term of parsed.excludedTerms) {
    if (matchesFuzzySearch(term, targets)) return false;
  }

  if (parsed.bpm && !matchesNumberRange(sample.bpm, parsed.bpm)) return false;
  if (parsed.playbackType && normalizePlaybackLabel(sample.playback_type) !== parsed.playbackType && normalizePlaybackLabel(sample.sample_type) !== parsed.playbackType) return false;
  if (parsed.instrumentType && normalizeValue(sample.instrument_type) !== parsed.instrumentType) return false;
  if (parsed.key && normalizeKey(sample.musical_key ?? "") !== parsed.key) return false;
  if (parsed.favorite !== null && isFavorite !== parsed.favorite) return false;

  for (const tag of parsed.tags) {
    if (!sample.tags.some((sampleTag) => normalizeValue(sampleTag) === tag)) return false;
  }

  for (const tag of parsed.excludedTags) {
    if (sample.tags.some((sampleTag) => normalizeValue(sampleTag) === tag)) return false;
  }

  return true;
}

function tokenize(query: string) {
  const tokens: string[] = [];
  let current = "";
  let inQuotes = false;

  for (const char of query) {
    if (char === '"') {
      inQuotes = !inQuotes;
    } else if (/\s/u.test(char) && !inQuotes) {
      if (current) {
        tokens.push(current);
        current = "";
      }
    } else {
      current += char;
    }
  }

  if (current) tokens.push(current);
  return tokens;
}

function pushText(parsed: SampleSearchDsl, negated: boolean, value: string) {
  const normalized = normalizeValue(value);
  if (!normalized) return;
  if (negated) parsed.excludedTerms.push(normalized);
  else parsed.textTerms.push(normalized);
}

function pushTag(parsed: SampleSearchDsl, negated: boolean, value: string) {
  const normalized = normalizeValue(value);
  if (!normalized) return;
  if (negated) parsed.excludedTags.push(normalized);
  else parsed.tags.push(normalized);
}

function parseNumberRange(value: string): NumberRange | null {
  const separator = value.indexOf("-");
  if (separator !== -1) {
    const range = {
      min: parseNumber(value.slice(0, separator)),
      max: parseNumber(value.slice(separator + 1)),
    };
    return range.min !== null || range.max !== null ? range : null;
  }

  const number = parseNumber(value);
  return number === null ? null : { min: number, max: number };
}

function parseNumber(value: string) {
  const trimmed = value.trim();
  if (!trimmed) return null;
  const parsed = Number.parseFloat(trimmed);
  return Number.isFinite(parsed) ? parsed : null;
}

function matchesNumberRange(value: number | null, range: NumberRange) {
  if (value === null) return false;
  if (range.min !== null && value < range.min) return false;
  if (range.max !== null && value > range.max) return false;
  return true;
}

function parseBoolean(value: string) {
  const normalized = normalizeValue(value);
  if (["true", "yes", "1"].includes(normalized)) return true;
  if (["false", "no", "0"].includes(normalized)) return false;
  return null;
}

function normalizeField(value: string) {
  return normalizeValue(value).replace(/[_-]/gu, "");
}

function normalizeValue(value: string) {
  return value.normalize("NFKC").trim().toLowerCase();
}

function normalizePlaybackType(value: string) {
  const normalized = normalizePlaybackLabel(value);
  return normalized === "loop" || normalized === "oneshot" ? normalized : null;
}

function normalizePlaybackLabel(value: string) {
  const normalized = normalizeValue(value).replace(/[_\s-]/gu, "");
  if (normalized === "one" || normalized === "shot" || normalized === "oneshot") return "oneshot";
  return normalized;
}

function normalizeKey(value: string) {
  const normalized = normalizeValue(value);
  const withoutMinor = normalized.endsWith("m") ? normalized.slice(0, -1) : normalized;
  if (!withoutMinor) return null;

  const root = withoutMinor[0].toUpperCase();
  const suffix = withoutMinor.slice(1);
  const key = suffix === "b" ? normalizeFlatKey(root) : `${root}${suffix}`;
  const allowed = ["C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B"];
  return key && allowed.includes(key) ? key : null;
}

function normalizeFlatKey(root: string) {
  const flats: Record<string, string> = {
    D: "C#",
    E: "D#",
    G: "F#",
    A: "G#",
    B: "A#",
  };
  return flats[root] ?? null;
}

File: core/src/db/operations/search_dsl.rs
use super::fuzzy::matches_fuzzy_query;
use super::types::SampleRow;

#[derive(Debug, Clone, PartialEq)]
pub(super) struct SampleSearchQuery {
    text_terms: Vec<String>,
    excluded_terms: Vec<String>,
    bpm: Option<NumberRange>,
    playback_type: Option<String>,
    instrument_type: Option<String>,
    key: Option<String>,
    tags: Vec<String>,
    excluded_tags: Vec<String>,
}

#[derive(Debug, Clone, Copy, PartialEq)]
struct NumberRange {
    min: Option<f64>,
    max: Option<f64>,
}

impl SampleSearchQuery {
    pub(super) fn parse(query: &str) -> Self {
        let mut parsed = Self {
            text_terms: Vec::new(),
            excluded_terms: Vec::new(),
            bpm: None,
            playback_type: None,
            instrument_type: None,
            key: None,
            tags: Vec::new(),
            excluded_tags: Vec::new(),
        };

        let tokens = tokenize(query);
        if tokens.is_empty() && !query.trim().is_empty() {
            parsed.push_text(false, query);
            return parsed;
        }

        for token in tokens {
            let (negated, body) = match token.strip_prefix('-') {
                Some(rest) if !rest.is_empty() => (true, rest),
                _ => (false, token.as_str()),
            };

            let Some((raw_field, raw_value)) = body.split_once(':') else {
                parsed.push_text(negated, body);
                continue;
            };

            let field = normalize_field(raw_field);
            let value = raw_value.trim();
            if value.is_empty() {
                parsed.push_text(negated, body);
                continue;
            }

            match field.as_str() {
                "bpm" => {
                    if !negated {
                        parsed.bpm = parse_number_range(value);
                    }
                }
                "type" | "playback" | "playbacktype" | "sampletype" => {
                    if !negated {
                        parsed.playback_type = normalize_playback_type(value);
                    }
                }
                "instrument" | "instrumenttype" => {
                    if !negated {
                        parsed.instrument_type = Some(normalize_value(value));
                    }
                }
                "key" => {
                    if !negated {
                        parsed.key = normalize_key(value);
                    }
                }
                "tag" | "tags" => parsed.push_tag(negated, value),
                "favorite" | "favourite" | "fav" => {}
                _ => parsed.push_text(negated, body),
            }
        }

        parsed
    }

    pub(super) fn matches(&self, sample: &SampleRow) -> bool {
        let tag_text = sample.tags.join(" ");
        let text_targets = [sample.file_name.as_str(), tag_text.as_str()];

        for term in &self.text_terms {
            if !matches_fuzzy_query(term, &text_targets) {
                return false;
            }
        }

        for term in &self.excluded_terms {
            if matches_fuzzy_query(term, &text_targets) {
                return false;
            }
        }

        if let Some(range) = self.bpm {
            if !matches_number_range(sample.bpm, range) {
                return false;
            }
        }

        if let Some(playback_type) = &self.playback_type {
            if normalize_playback_label(&sample.playback_type) != *playback_type
                && sample
                    .sample_type
                    .as_deref()
                    .map(normalize_playback_label)
                    .as_deref()
                    != Some(playback_type.as_str())
            {
                return false;
            }
        }

        if let Some(instrument_type) = &self.instrument_type {
            if normalize_value(&sample.instrument_type) != *instrument_type {
                return false;
            }
        }

        if let Some(key) = &self.key {
            if sample
                .musical_key
                .as_deref()
                .and_then(normalize_key)
                .as_deref()
                != Some(key)
            {
                return false;
            }
        }

        for tag in &self.tags {
            if !sample
                .tags
                .iter()
                .any(|sample_tag| normalize_value(sample_tag) == *tag)
            {
                return false;
            }
        }

        for tag in &self.excluded_tags {
            if sample
                .tags
                .iter()
                .any(|sample_tag| normalize_value(sample_tag) == *tag)
            {
                return false;
            }
        }

        true
    }

    fn push_text(&mut self, negated: bool, value: &str) {
        let normalized = normalize_value(value);
        if normalized.is_empty() {
            return;
        }
        if negated {
            self.excluded_terms.push(normalized);
        } else {
            self.text_terms.push(normalized);
        }
    }

    fn push_tag(&mut self, negated: bool, value: &str) {
        let normalized = normalize_value(value);
        if normalized.is_empty() {
            return;
        }
        if negated {
            self.excluded_tags.push(normalized);
        } else {
            self.tags.push(normalized);
        }
    }
}

fn tokenize(query: &str) -> Vec<String> {
    let mut tokens = Vec::new();
    let mut current = String::new();
    let mut in_quotes = false;

    for c in query.chars() {
        match c {
            '"' => in_quotes = !in_quotes,
            c if c.is_whitespace() && !in_quotes => {
                if !current.is_empty() {
                    tokens.push(std::mem::take(&mut current));
                }
            }
            _ => current.push(c),
        }
    }

    if !current.is_empty() {
        tokens.push(current);
    }

    tokens
}

fn parse_number_range(value: &str) -> Option<NumberRange> {
    if let Some((min, max)) = value.split_once('-') {
        let range = NumberRange {
            min: parse_number(min),
            max: parse_number(max),
        };
        return (range.min.is_some() || range.max.is_some()).then_some(range);
    }

    parse_number(value).map(|number| NumberRange {
        min: Some(number),
        max: Some(number),
    })
}

fn parse_number(value: &str) -> Option<f64> {
    let trimmed = value.trim();
    if trimmed.is_empty() {
        return None;
    }
    trimmed
        .parse::<f64>()
        .ok()
        .filter(|number| number.is_finite())
}

fn matches_number_range(value: Option<f64>, range: NumberRange) -> bool {
    let Some(value) = value else {
        return false;
    };
    if let Some(min) = range.min {
        if value < min {
            return false;
        }
    }
    if let Some(max) = range.max {
        if value > max {
            return false;
        }
    }
    true
}

fn normalize_field(value: &str) -> String {
    normalize_value(value).replace(['_', '-'], "")
}

fn normalize_value(value: &str) -> String {
    value
        .chars()
        .flat_map(|c| match c {
            '\u{3000}' => ' '.to_lowercase(),
            '\u{ff01}'..='\u{ff5e}' => {
                let ascii = char::from_u32(c as u32 - 0xfee0).unwrap_or(c);
                ascii.to_lowercase()
            }
            _ => c.to_lowercase(),
        })
        .collect::<String>()
        .trim()
        .to_string()
}

fn normalize_playback_type(value: &str) -> Option<String> {
    let normalized = normalize_playback_label(value);
    match normalized.as_str() {
        "loop" | "oneshot" => Some(normalized),
        _ => None,
    }
}

fn normalize_playback_label(value: &str) -> String {
    match normalize_value(value).replace(['_', '-', ' '], "").as_str() {
        "one" | "shot" | "oneshot" => "oneshot".to_string(),
        other => other.to_string(),
    }
}

fn normalize_key(value: &str) -> Option<String> {
    let normalized = normalize_value(value);
    let without_minor = normalized.strip_suffix('m').unwrap_or(&normalized);
    let mut chars = without_minor.chars();
    let root = chars.next()?.to_ascii_uppercase();
    let suffix = chars.collect::<String>();
    let key = match suffix.as_str() {
        "" => root.to_string(),
        "#" => format!("{root}#"),
        "b" => normalize_flat_key(root)?,
        _ => return None,
    };

    matches!(
        key.as_str(),
        "C" | "C#" | "D" | "D#" | "E" | "F" | "F#" | "G" | "G#" | "A" | "A#" | "B"
    )
    .then_some(key)
}

fn normalize_flat_key(root: char) -> Option<String> {
    match root {
        'D' => Some("C#".to_string()),
        'E' => Some("D#".to_string()),
        'G' => Some("F#".to_string()),
        'A' => Some("G#".to_string()),
        'B' => Some("A#".to_string()),
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn sample() -> SampleRow {
        SampleRow {
            id: 1,
            path: "/samples/kick.wav".to_string(),
            file_name: "metal_kick.wav".to_string(),
            duration: Some(1.0),
            bpm: Some(140.0),
            periodicity: Some(0.8),
            sample_rate: Some(44_100),
            file_size: None,
            artist: None,
            low_ratio: None,
            attack_slope: None,
            decay_time: None,
            sample_type: Some("oneshot".to_string()),
            waveform_peaks: None,
            embedding: None,
            is_online: true,
            playback_type: "oneshot".to_string(),
            instrument_type: "kick".to_string(),
            musical_key: Some("A".to_string()),
            content_hash: None,
            duplicate_count: 1,
            tags: vec!["metal".to_string(), "drums".to_string()],
        }
    }

    #[test]
    fn parses_and_matches_sample_dsl() {
        let query = SampleSearchQuery::parse("kick bpm:120-180 type:one-shot tag:metal key:Am");

        assert!(query.matches(&sample()));
    }

    #[test]
    fn excludes_negative_text_and_tags() {
        assert!(!SampleSearchQuery::parse("kick -metal").matches(&sample()));
        assert!(!SampleSearchQuery::parse("kick -tag:metal").matches(&sample()));
    }

    #[test]
    fn ignores_frontend_only_favorite_clause() {
        assert!(SampleSearchQuery::parse("favorite:true").matches(&sample()));
    }
}
